Source/CCZ4/CCZ4Geometry.hpp

```diff
@@ -107,6 +109,62 @@ class CCZ4Geometry
         return out;
     }
 
+    template <class data_t>
+    static Tensor<2, data_t>
+    compute_d1_chris_contracted(const Tensor<2, data_t> &h_UU,
+                                const Tensor<2, Tensor<1, data_t>> &d1_h,
+                                const Tensor<2, Tensor<2, data_t>> &d2_h)
+    {
+        Tensor<2, data_t> d1_chris_contracted = 0.0;
+        FOR2(i, j)
+        {
+            FOR3(m, n, p)
+            {
+                data_t d1_terms = 0.0;
+                FOR2(q, r)
+                {
+                    d1_terms += -h_UU[q][r] * (d1_h[n][q][j] * d1_h[m][p][r] +
+                                               d1_h[m][n][j] * d1_h[p][q][r]);
+                }
+                d1_chris_contracted[i][j] +=
+                    h_UU[i][m] * h_UU[n][p] * (d2_h[m][n][j][p] + d1_terms);
+            }
+        }
+        return d1_chris_contracted;
+    }
+
+    // This function allows adding arbitrary multiples of D_{(i}Z_{j)}
+    // to the Ricci scalar rather than the default of 2 in compute_ricci_Z
+    template <class data_t, template <typename> class vars_t,
+              template <typename> class diff2_vars_t>
+    static ricci_t<data_t>
+    compute_ricci_Z_general(const vars_t<data_t> &vars,
+                            const vars_t<Tensor<1, data_t>> &d1,
+                            const diff2_vars_t<Tensor<2, data_t>> &d2,
+                            const Tensor<2, data_t> &h_UU,
+                            const chris_t<data_t> &chris, const double dZ_coeff)
+    {
+        // get contributions from conformal metric and factor with zero Z vector
+        Tensor<1, data_t> Z0 = 0.;
+        auto ricci = compute_ricci_Z(vars, d1, d2, h_UU, chris, Z0);
+
+        // need to add term to correct for d1.Gamma as this has contribution
+        // from Z
+        auto d1_chris_contracted =
+            compute_d1_chris_contracted(h_UU, d1.h, d2.h);
+        FOR3(i, j, m)
+        {
+            ricci.LL[i][j] +=
+                (1. - 0.5 * dZ_coeff) * 0.5 *
+                (vars.h[m][i] * (d1_chris_contracted[m][j] - d1.Gamma[m][j]) +
+                 vars.h[m][j] * (d1_chris_contracted[m][i] - d1.Gamma[m][i]));
```

Comment:
I don't think `Z_over_chi` needs to be passed. It can just be calculated as `0.5 * (vars.Gamma - chris.contracted)`. I have checked your expressions and I think the terms after the 2nd `+=` should be divided by `chi` (note that `z_terms` is divided by `chi` in L104 at commit c58a6a2). Would you be able to double check this?

Reply:
True true true. Divide by chi yes!

Reply:
and the last `0.5` on L158 shouldn't be multiplying the terms in L161.

Reply:
Done.